### PR TITLE
os/bluestore: Track bluestore ops

### PR DESCRIPTION
**Pull request by yanghonggang**
Signed-off-by: Yang Honggang <yanghonggang_yewu@cmss.chinamobile.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by yanghonggang** [comment]
jenkins test make check

**Comment by ifed01** [comment]
I'm not familiar with TrackedOp stuff but given additional code coming from https://github.com/ceph/ceph/pull/41719 I would suggest to isolate all the tracking stuff at either TrackedOp itself or at an additional single helper object dispatching notifications to all the targets: TrackedOp, Blkin, Jaeger and whatever other tool we introduce in the future. Without that bluestore code becomes hardly readable with all these ifdef-s, if-s and  notification calls....

**Comment by yanghonggang** [comment]
> I'm not familiar with TrackedOp stuff but given additional code coming from #41719 I would suggest to isolate all the tracking stuff at either TrackedOp itself or at an additional single helper object dispatching notifications to all the targets: TrackedOp, Blkin, Jaeger and whatever other tool we introduce in the future. Without that bluestore code becomes hardly readable with all these ifdef-s, if-s and notification calls....

you are right. I will isolate TracedOp, Blkin and Jaeger at TrackedOp.

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -11250,6 +11250,10 @@ BlueStore::TransContext *BlueStore::_txc_create(
 {
   TransContext *txc = new TransContext(cct, c, osr, on_commits);
   txc->t = db->get_transaction();
+  txc->osd_op = osd_op;
+
+  if (osd_op)
+    osd_op->mark_event("txc create");
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.cc:11256]: code style: even single statements with { }.

File: src/os/bluestore/BlueStore.h [hunk 2]
```diff
@@ -1563,6 +1563,10 @@ class BlueStore : public ObjectStore,
 
     inline void set_state(state_t s) {
        state = s;
+
+       if (osd_op)
+         osd_op->mark_event(get_state_name());
```
Comment (aclamk) [inline src/os/bluestore/BlueStore.h:1568]: Adding this code changed set_state from perfectly inlined to separate function.
While this is not your fault, and overall impact is still acceptable, how about following:
1) add "unlikely" to condition
2) add get_state_name variant that retuerns std:string_view to. Problem with conversion from const char* to string_view is that length() has to be invoked.
These are of course optional and do not affect PR's usefulness.

File: src/os/bluestore/BlueStore.cc (hunk 1 above)
Comment (yanghonggang) [reply src/os/bluestore/BlueStore.cc:11256]: ok

File: src/os/bluestore/BlueStore.h (hunk 2 above)
Comment (yanghonggang) [reply src/os/bluestore/BlueStore.h:1568]: OK

**Comment by yanghonggang** [comment]
> I'm not familiar with TrackedOp stuff but given additional code coming from #41719 I would suggest to isolate all the tracking stuff at either TrackedOp itself or at an additional single helper object dispatching notifications to all the targets: TrackedOp, Blkin, Jaeger and whatever other tool we introduce in the future. Without that bluestore code becomes hardly readable with all these ifdef-s, if-s and notification calls....

you are right. And as we will have Jaeger now, the Blkin code in bluestore maybe can be cleaned up now. What's your suggestions? @ifed01 @ideepika

**Comment by ifed01** [comment]
> > I'm not familiar with TrackedOp stuff but given additional code coming from #41719 I would suggest to isolate all the tracking stuff at either TrackedOp itself or at an additional single helper object dispatching notifications to all the targets: TrackedOp, Blkin, Jaeger and whatever other tool we introduce in the future. Without that bluestore code becomes hardly readable with all these ifdef-s, if-s and notification calls....
> 
> you are right. And as we will have Jaeger now, the Blkin code in bluestore maybe can be cleaned up now. What's your suggestions? @ifed01 @ideepika

No idea, sorry...

**Comment by ideepika** [comment]
> > I'm not familiar with TrackedOp stuff but given additional code coming from #41719 I would suggest to isolate all the tracking stuff at either TrackedOp itself or at an additional single helper object dispatching notifications to all the targets: TrackedOp, Blkin, Jaeger and whatever other tool we introduce in the future. Without that bluestore code becomes hardly readable with all these ifdef-s, if-s and notification calls....
> 
> you are right. And as we will have Jaeger now, the Blkin code in bluestore maybe can be cleaned up now. What's your suggestions? @ifed01 @ideepika

yes, we plan on moving to Jaeger  completely, should be fine to clean up Blkin stuff

**Comment by yanghonggang** [comment]
> > > I'm not familiar with TrackedOp stuff but given additional code coming from #41719 I would suggest to isolate all the tracking stuff at either TrackedOp itself or at an additional single helper object dispatching notifications to all the targets: TrackedOp, Blkin, Jaeger and whatever other tool we introduce in the future. Without that bluestore code becomes hardly readable with all these ifdef-s, if-s and notification calls....
> > 
> > 
> > you are right. And as we will have Jaeger now, the Blkin code in bluestore maybe can be cleaned up now. What's your suggestions? @ifed01 @ideepika
> 
> @yangdongsheng yes, we plan on moving to Jaeger completely, should be fine to clean up Blkin stuff

OK

File: src/os/bluestore/BlueStore.h (hunk 2 above)
Comment (yanghonggang) [reply src/os/bluestore/BlueStore.h:1568]: >     1. unlikely

@aclamk 
For the following code, there will be 'error: cannot convert'.

```cpp
    inline void set_state(state_t s) {
       state = s;
       if (unlikely(tracer)) {
         tracer->mark_tracepoint(get_state_name());
       }
    }
```

```text
/home/hehe/work/gits/ceph.yhg/src/os/bluestore/BlueStore.h: In member function ‘void BlueStore::TransContext::set_state(BlueStore::TransContext::state_t)’:
/home/hehe/work/gits/ceph.yhg/src/os/bluestore/BlueStore.h:1562:21: error: cannot convert ‘TrackedOpRef’ {aka ‘boost::intrusive_ptr<TrackedOp>’} to ‘long int’
 1562 |        if (unlikely(tracer)) {
/home/hehe/work/gits/ceph.yhg/src/common/likely.h:25:43: note: in definition of macro ‘unlikely’
   25 | #define unlikely(x)     __builtin_expect((x),0)
```

**Pull request closed without merging**